File: JavaScript/DoublyLinkedList.js
// it's a list that contains links to previous and next nodes

// very similar to single link list. twice more efficient for access but takes twice more extra storage

class Node {
    constructor(val, prev=null, next=null) {
        this.prev = prev;
        this.val = val;
        this.next = next;
    }
}

class DoublyLinkedList {
    constructor(){
        this.head = null;
        this.tail = null;
        this.length = 0;
    }
    // push
    push(itm) {
        let newNode = new Node(itm)
        if (!this.head) {
            this.head = newNode;
            this.tail = newNode;
        }
        else {
            newNode.prev = this.tail;
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
    }
    // pop
    pop() {
        if (!this.head) return undefined;
        let poppedItem = this.tail;
        if (this.length === 1) {
            this.clear();
        }
        else {
            this.tail = this.tail.prev;
            this.tail.next = null;
        }
        this.length--;
        return poppedItem.val;
    }
    // shift
    shift() {
        if (!this.head) return undefined;
        let shiftedItem = this.head;
        if (this.length === 1) {
            this.clear();
            return shiftedItem.val;
        }
        this.head = this.head.next;
        this.head.prev = null;
        this.length--;
        return shiftedItem.val;
    }
    // unshift
    unshift(itm) {
        let newNode = new Node(itm);
        if (!this.head) {
            this.head = newNode;
            this.tail = newNode;
        }
        else {
            newNode.next = this.head;
            this.head.prev = newNode;
            this.head = newNode;
        }
        this.length++;
    }
    // get
    get(idx) {
        if (idx <= 0) return this.head.val;
        if (idx >= this.length) return this.tail.val;

        let current = this.traverse(idx);
        return current.val;
    }
    // set
    set(idx, itm) {
        if (idx < 0) return this.unshift(itm);
        if (idx >= this.length) return this.push(itm);

        let current = this.traverse(idx);
        current.val = itm;
    }
    // insert
    insert(idx, itm) {
        if (idx <= 0) return this.unshift(itm);
        if (idx >= this.length) return this.push(itm);

        let newNode = new Node(itm);
        let current = this.traverse(idx);
        newNode.prev = current.prev;
        current.prev.next = newNode;
        current.prev = newNode;
        newNode.next = current;
        current = newNode;
        this.length++;
    }
    // remove
    remove(idx) {
        if (idx <= 0) return this.shift();
        if (idx >= this.length - 1) return this.pop();

        let current = this.traverse(idx);
        let removedItem = current;
        current.prev.next = current.next;
        current.next.prev = current.prev;
        this.length--;
        return removedItem.val;
    }
    // reverse
    reverse() {
        if (this.length <= 1) return;

        let current = this.head;
        for (let i = 0; i < this.length; i++) {
            [current.prev, current.next] = [current.next, current.prev];
            current = current.prev;
        }
        [this.head, this.tail] = [this.tail, this.head];
    }
    // print all elements
    print() {
        if (!this.head) return undefined;
        let current = this.head;
        let i = 0;
        while (current) {
            console.log(`[${i}]`, current.val);
            current = current.next;
            i++;
        }
    }
    // empty list
    clear() {
        if (!this.head) return undefined;
        this.head = null;
        this.tail = null;
        this.length = 0;
    }
    // traverse to index
    traverse(idx) {
        let itm = null;
        if (idx < Math.floor(this.length / 2)) {
            // index in left half, traverse from beginning
            itm = this.head;
            for (let i = 0; i < idx; i++) {
                itm = itm.next;
            }
        }
        else {
            // index in right half, traverse from end
            itm = this.tail;
            for (let i = 1; i < this.length - idx; i++) {
                itm = itm.prev;
            }
        }
        return itm;
    }
}

// test drive code
let list = new DoublyLinkedList();
for (let i = 0; i < 14; i++) {
    list.push(i + 100);
}
console.log("Before: ")
list.print();

// test commands go here
// console.log("Pop: ", list.pop());
// console.log("Shift: ", list.shift());
// list.unshift(99);
// console.log("Get: ", list.get(39));
// list.set(99, 1);
// list.insert(2, "----");
// console.log("Remove: ", list.remove(2));
// list.reverse();

// console.log("After:")
// list.print();

File: GoLang/model/doublyLinkedList.go
package model

import (
	"errors"
	"fmt"
)

type DIntNode struct {
	Previous *DIntNode
	Value    int
	Next     *DIntNode
}

type DIntLinkedList struct {
	Head   *DIntNode
	Tail   *DIntNode
	Length int
}

func (dl *DIntLinkedList) GetLength() int {
	return dl.Length
}

func (dl *DIntLinkedList) Print() {
	fmt.Printf("length: %d --> [", dl.GetLength())
	if dl.Head == nil {
		return
	}
	fmt.Printf("%d, ", dl.Head.Value)
	next := dl.Head.Next
	for next != nil {
		fmt.Printf("%d, ", next.Value)
		next = next.Next
	}
	fmt.Println("]")
}

func (dl *DIntLinkedList) Push(x int) {
	if dl.Length == 0 {
		node := &DIntNode{Value: x}
		dl.Head = node
		dl.Tail = node
	} else {
		node := &DIntNode{Value: x, Previous: dl.Tail}
		dl.Tail.Next = node
		dl.Tail = node
	}
	dl.Length++
}

func (dl *DIntLinkedList) Pop() (int, bool) {
	if dl.Length <= 0 {
		return 0, false
	} else if dl.Length == 1 {
		val := dl.Tail.Value
		dl.Head = nil
		dl.Tail = nil
		dl.Length--
		return val, true
	}
	node := dl.Tail
	dl.Tail = dl.Tail.Previous
	dl.Tail.Next = nil
	dl.Length--
	return node.Value, true
}

func (dl *DIntLinkedList) Shift() (int, bool) {
	if dl.Length <= 0 {
		return 0, false
	} else if dl.Length == 1 {
		val := dl.Head.Value
		dl.Head = nil
		dl.Tail = nil
		dl.Length--
		return val, true
	}
	val := dl.Head.Value
	dl.Head = dl.Head.Next
	dl.Head.Previous = nil
	return val, true
}

func (dl *DIntLinkedList) Unshift(x int) {
	node := &DIntNode{Value: x}
	if dl.Length <= 0 {
		dl.Head = node
		dl.Tail = node
	}
	node.Next = dl.Head
	dl.Head.Previous = node
	dl.Head = node
	dl.Length++
}

func (dl *DIntLinkedList) Get(i int) (int, bool) {
	if i < 0 {
		return 0, false
	} else if i >= dl.Length {
		return 0, false
	}

	start := dl.Head
	j := 0
	for j < i {
		start = start.Next
		j++
	}
	return start.Value, true
}

func (dl *DIntLinkedList) Set(pos, val int) error {
	if pos < 0 {
		return errors.New("index out of bounds")
	} else if pos == 0 {
		dl.Head.Value = val
	} else if pos >= dl.GetLength() {
		return errors.New("index out of bounds")
	} else {
		node := dl.Head
		i := 0
		for i < pos {
			node = node.Next
			i++
		}
		node.Value = val
	}
	return nil
}

func (dl *DIntLinkedList) Insert(pos, val int) {
	if pos <= 0 {
		dl.Unshift(val)
	} else if pos >= dl.GetLength()-1 {
		dl.Push(val)
	} else {
		curr := dl.Head
		i := 0
		for i < pos {
			curr = curr.Next
			i++
		}
		newNode := &DIntNode{Value: val, Next: curr, Previous: curr.Previous}
		curr.Previous.Next = newNode
		curr = newNode
	}
	dl.Length++
}

func (dl *DIntLinkedList) Remove(pos int) (int, bool) {
	if pos < 0 {
		return 0, false
	} else if pos == 0 {
		return dl.Shift()
	} else if pos >= dl.GetLength() {
		return 0, false
	} else if pos == dl.GetLength()-1 {
		return dl.Pop()
	}
	i := 0
	curr := dl.Head
	for i < pos {
		curr = curr.Next
		i++
	}
	tempNode := curr
	curr.Previous.Next = curr.Next
	curr.Next.Previous = curr.Previous
	dl.Length--
	return tempNode.Value, true
}

func (dl *DIntLinkedList) Reverse() {
	if dl.Length <= 1 {
		return
	}
	node := dl.Head
	dl.Head = dl.Tail
	dl.Tail = node

	var next *DIntNode
	var prev *DIntNode

	for i := 0; i < dl.Length; i++ {
		next = node.Next
		node.Next = prev
		node.Previous = next
		node = next
	}
}
